### DISCUSSION TITLE
Generic Type type in form of alternitive generics

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [dotnet/csharplang#5010] [Language Ideas]
I apologize in advance if this issue seems simplistic.

# Generic `Type` type.
I recently ran into a issue when creating a unity style entity component system `as in EC system` not `ECS system`
Anyways when creating a required attribute for components I ran into the issue of attributes not supporting generic types.

To my surprise there is a proposal about it, however when I really thought about it the suggested solution didn't really solve my  issue.
For example..
- Generic Types don't support params
- Even if so there doesn't seem to be a way to store them without down casting them to Type

I had the idea instead of generic parameters what if we could pass generic `Type` types?
```cs
    public class Required : Attribute
    {
        public List<Component> components = new();  
        
        public Required(params Type<Component>[] components)
        {
           _ components.AddRange(components);
        }
    }

   //All derive from component.
   [Required(typeof(Transform), typeof(Physicsbody), typeof(Renderer), typeof(Collider)]
   public MyEntityTemplate : Entity { }
```

The generic `Type` type would be sealed but would also inherit from the generic type passed.

I am not sure if this would be compile time safe so someone would have to chime in on this, 
but the idea is that it would be.

**Comment by HaloFour** [comment]
This is probably more a suggestion for the [runtime](https://github.com/dotnet/runtime/issues/) repository which would have to define such a type.

`Class<T>` is useful in Java but almost all of the use cases revolve around plugging the lack of reification of generic type arguments in that language.  IMO it would be a lot less useful in .NET outside of a handful of very specific scenarios.

**Comment by Joe4evr** [comment]
Another problem with this proposed type is that, ***from your own example***, in order for `Type<Transform>` to be assignable to `Type<Component>` it'd require `Type<T>` to be covariant, which is not supported for classes.

**Reply by Shadowblitz16** [reply]
I didn't realize that.
Is there a reason why generics don't support covariant classes?

**Reply by Joe4evr** [reply]
The main complication is that fields can't work with variance. Imagine if the following *was* allowed:
```cs
public class Pen<out TAnimal>
    where TAnimal : Animal
{
    public TAnimal Occupant;
}
/*********/
void M()
{
    var pen = new Pen<Giraffe>() { Occupant = new Giraffe() };
    N(pen);
}

void N(Pen<Animal> pen)
{
    pen.Occupant = new Turtle();
}
```
The instance created in `M` has the field typed as `Giraffe`, which means it's impossible to fit an object of type `Turtle` in that field, but in `N` that information isn't known, so this would compile but blow up at runtime.

Of note, this Type System hole already exists with arrays since .NET 1.0 (with no real good way to fix it).
```cs
void M()
{
    var giraffes = new Giraffe[] { new Giraffe() };
    N(giraffes);
}

void N(Animal[] animals)
{
    animals[0] = new Turtle();
}
```
This actually *will* compile, [but will throw an `ArrayTypeMismatchException` at runtime](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfANwEMoACAcwEsoqAzVmAZ1oF5a8YAd1oBxZmw4BtALq0A3vyGjx7GAAoAlLQC+AbgCwAKAByapi1WcNBw0buGAAgEYktBylqmAgnkYBbKgAbGVoqXwDAqyM5I1o40PCgzkkABlk+AWEAFRwoDED1ayNte3sHAGZQ4E4MFjAMNwAmWh9/IPkdIwqqmrqGh2aAWSo/CNoQFsTAjpLHSqpq2qp6ptoAJRgABwxGAvHJtumFWe6B5QsOfeHR9uOuyrOcvL2Jje3dmBmjIA===).

**Reply by Shadowblitz16** [reply]
Odd I never thought of it that way.
It is a bit weird you can pass a Giraffe[] to a Animal[]. 
Is there any plans on fixing this in C#?

**Reply by CyrusNajmabadi** [reply]
> Is there any plans on fixing this in C#?

No. It would break the universe. If it's something you don't like, you can always write an analyzer to disallow it in your projects.

**Comment by PathogenDavid** [comment]
Is there a reason you aren't just using multiple attributes -- one for each component? ([SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQHMIDsIIJYDGqsM+wcMKANDCAG655gCeAPgAIBMAjALABQAgNoAlCAEc4eJAGEA9gFsADnJxYYAHgAqCKFmRg5CBQD4AusLGTpEectXZNYrABNcuc5YlTZilWs15ABsgvDcET352AGYAAi5YgFlmAFF1PBhmAQBvAQBfAWEAQRhSPHJKAFVkKEwAChKyiogtKARMGGQAOhkgqGRkKliikLkAd0S4IJg8JSCIWIBeWNI4CABKCyi45Gh5l3jOWKsfWz8HdUayCghtE1iQYdLrygFY99ixgAt3WK0H2J2fyOHKxAqCbaxKDAZCkKAEGCHQHnAKxbJggQxJE6PQGIwKAFAi6I9HgrEJZzhXCElGONEYyEJYKhcI0+yo0kCIA===))

```csharp
[RequireComponent<Transform>]
[RequireComponent<Renderer>]
[RequireComponent<Collider>]
public class MyEntity
{
}

[AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
public sealed class RequireComponentAttribute<T> : Attribute
    where T : Component
{ }
```

**Reply by Joe4evr** [reply]
Well, for one thing, generic attributes aren't shipped into a proper version of the compiler yet (not even the latest preview SDK). 😜

**Reply by PathogenDavid** [reply]
The subtext of this discussion involves generic attributes. Also, see https://github.com/dotnet/csharplang/issues/124#issuecomment-894498841 for more background context.
